### DISCUSSION TITLE
How to apply preamplifier response function to raw simulation pulses in SolidStateDetectors

### DISCUSSION THREAD

**Discussion by zqiuli** [JuliaPhysics/SolidStateDetectors.jl#360] [Q&A] [answered]
Dear Sir, I have already simulated the raw waveform as shown below. Now I want to **apply preamplifier response function** to them, I found **simulate_waveforms** can't do this. 
**Quesion1**. Is there any other method ? If available, I also would like to save the convolved waveform in a file.
![image](https://github.com/JuliaPhysics/SolidStateDetectors.jl/assets/145674078/9d67f26b-5d45-4c8c-8c06-6a0fb89f9d89)
**Quesion2**. I have seen the source code for the **charge cloud**  in src/ChargeCloudModels/ However, I still don't know how to **set the parameters "number_of_carriers" and "number_of_shells"** in function **simulate_waveforms** when considering charge cloud. If you could explain how SolidStateDetectors consider charge clouds in simple terms(or recommand some relative references) and provide guidance on setting "number_of_carriers" and "number_of_shells," I would greatly appreciate it.

**Quesion3.** The code in src/ChargeCloudModels/PlatonicSolids/PlatonicSolids.jl is below
In the case of Hexahedron, it has 6 faces, why define **const Hexahedron{T} = PlatonicSolid{8, T}** instead of const Hexahedron{T} = PlatonicSolid{6, T}. 
`const Hexahedron{T} = PlatonicSolid{8, T}
function Hexahedron(center::CartesianPoint{T}, length::Real = T(1))::Hexahedron{T} where {T}
    structure::NTuple{4, Tuple{Int,T,T}} = ((1,T(1),T(0)),(3,T(1/3),T(0)),(3,T(-1/3),T(π/3)),(1,T(-1),T(0)))
    PlatonicSolid(get_shell_structure_locations(structure, center, T(length)))
end

const Octahedron{T} = PlatonicSolid{6, T}
function Octahedron(center::CartesianPoint{T}, length::Real = T(1))::Octahedron{T} where {T}
    structure::NTuple{3, Tuple{Int,T,T}} = ((1,T(1),T(0)), (4,T(0),T(0)), (1,T(-1),T(0)))
    PlatonicSolid(get_shell_structure_locations(structure, center, T(length)))
end

const Icosahedron{T} = PlatonicSolid{12, T}
function Icosahedron(center::CartesianPoint{T}, length::Real = T(1))::Icosahedron{T} where {T}
    structure::NTuple{4, Tuple{Int,T,T}} = ((1,T(1),T(0)), (5,T(sqrt(1/5)),T(0)), (5,T(-sqrt(1/5)),T(π/5)), (1,T(-1),T(0)))
    PlatonicSolid(get_shell_structure_locations(structure, center, T(length)))
end

const Dodecahedron{T} = PlatonicSolid{20, T}
function Dodecahedron(center::CartesianPoint{T}, length::Real = T(1))::Dodecahedron{T} where {T}
    zmax::T = sqrt((5 + 2*sqrt(5))/15)
    tmp::T = 1 - 2*(1 - zmax^2) * sin(π/5)^2
    zmin::T = zmax * tmp - sqrt(1 - zmax^2) * sqrt(1 - tmp^2)
    structure::NTuple{4, Tuple{Int,T,T}} = ((5,zmax,T(0)),(5,zmin,T(0)),(5,-zmin,T(π/5)),(5,-zmax,T(π/5)))
    PlatonicSolid(get_shell_structure_locations(structure, center, T(length)))
end`
Look forward to your reply,thank you very much.

**Comment by fhagemann** [comment, marked as answer]
Hi,
1. applying response functions is **not** part of SSD, we stop at where the **raw** waveforms are simulated.
A package dedicated to apply filters to the output of SSD (`RDWaveforms`) is [RadiationDetectorDSP.jl](https://github.com/JuliaPhysics/RadiationDetectorDSP.jl), e.g. in [`circuit_filters.jl`](https://github.com/JuliaPhysics/RadiationDetectorDSP.jl/blob/main/src/circuit_filters.jl)
2. You need to think about two things when simulating charge clouds:
   a) the initial configuration of the charge carriers at the time of electron-hole-pair creation
  b)  charge cloud effects => diffusion and self-repulsion of the charge carriers
The keywords `number_of_carriers` and `number_of_shells` define the initial configuration. In `simulate_waveforms`, the charge carriers are evenly distributed on surfaces of spheres. `number_of_shells` gives the numbers of spheres, `number_of_carriers` the number of charge carriers created. 
For example, defining a charge cloud with `number_of_shells = 2` and `number_of_carriers = 100`, you will receive a charge cloud consisting of a center charge at the given position, and then two spheres with around 100 (the exact number will depend on `number_of_shells`, but it should be within 5% of the number you give it) charge carriers on them:
![image](https://github.com/JuliaPhysics/SolidStateDetectors.jl/assets/30291312/2d5c3441-ecbc-41e6-82f3-9608ade46aa5)
Then, in the waveform simulation, you need to set `diffusion = true` and `self_repulsion = true` to take the charge cloud effects into account.
3. We use the **edges** of the platonic solid to spawn the charge carriers, not the **faces**. A hexahedron has indeed 6 faces, but 8 edges. That's why we define it as `PlatonicSolid{8,T}`:
![104365868-d5d4c800-5518-11eb-8018-6454332a50d8](https://github.com/JuliaPhysics/SolidStateDetectors.jl/assets/30291312/45ed410a-98a9-4265-95e0-b02ff87b998a)

**Reply by zqiuli** [reply]
I understand it. Thank you for your detailed reply!
